File: programmers/72412.py
def solution(info, query):
    answer = []
    infos, queries = [], []
    for i in info:
        info_arr = i.split(' ')
        tmp = []
        for i in range(0, 4):
            tmp.append(info_arr[i][0])
        tmp.append(int(info_arr[-1]))
        infos.append(tmp)
    for q in query:
        query_arr = q.replace(' and', '').split(' ')
        tmp = []
        for i in range(0, 4):
            tmp.append(query_arr[i][0])
        tmp.append(int(query_arr[-1]))
        queries.append(tmp)
    
    for i in range(len(query)):
        cnt = 0
        for j in range(len(infos)):
            for k in range(len(queries[i][0:-1])):
                if infos[j][k] != queries[i][k] and queries[i][k] != '-':
                    break   # 다음사람
                if k == 3 and infos[j][-1] >= queries[i][-1]:
                    cnt += 1
        answer.append(cnt)
    # print(answer)
    return answer

solution(["java backend junior pizza 150",
    "python frontend senior chicken 210",
    "python frontend senior chicken 150",
    "cpp backend senior pizza 260",
    "java backend junior chicken 80",
    "python backend senior chicken 50"], 
    ["java and backend and junior and pizza 100",
    "python and frontend and senior and chicken 200",
    "cpp and - and senior and pizza 250",
    "- and backend and senior and - 150",
    "- and - and - and chicken 100",
    "- and - and - and - 150"])

File: programmers/72412.js
const infoMap = new Map();

function combination(s, score, start) {
	const key = s.join("");
	if (infoMap[key]) infoMap[key].push(+score);
	else infoMap[key] = [+score];

	for (let i = start; i < s.length; i++) {
		const t = [...s];
		t[i] = "-";
		combination(t, score, i + 1);
	}
}

function solution(info, query) {
	let answer = [];

	for (let i = 0; i < info.length; i++) {
		const s = info[i].split(" ");
		const score = s.pop();
		combination(s, score, 0);
	}

	for (const key in infoMap) infoMap[key].sort((a, b) => a - b);

	// query
	for (const q of query) {
		const splited = q.split(/ and | /g);
		const score = +splited.pop();
		const key = splited.join("");
		const arr = infoMap[key];

		if (!arr) {
			answer.push(0);
			continue;
		}

		let [left, right] = [0, arr.length];
		while (left < right) {
			let mid = Math.floor((left + right) / 2);
			if (score <= arr[mid]) {
				right = mid;
			} else if (arr[mid] < score) {
				left = mid + 1;
			}
		}
		const result = arr.length - left;
		answer.push(result);
	}

	return answer;
}

console.log(
	solution(
		[
			"java backend junior pizza 150",
			"python frontend senior chicken 210",
			"python frontend senior chicken 150",
			"cpp backend senior pizza 260",
			"java backend junior chicken 80",
			"python backend senior chicken 50",
		],
		[
			"java and backend and junior and pizza 100",
			"python and frontend and senior and chicken 200",
			"cpp and - and senior and pizza 250",
			"- and backend and senior and - 150",
			"- and - and - and chicken 100",
			"- and - and - and - 150",
		]
	)
);
